### PR TITLE
Jon/23/multiple cassandra versions

### PR DESCRIPTION
**Pull request by rustyrazorblade**
CASSANDRASC-23

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -100,10 +100,12 @@ jobs:
...
       - install_java:
           version: adoptopenjdk-11-hotspot
-
+      - restore_cache:
+          keys:
+            - tarballs
       - run: sudo update-java-alternatives -s adoptopenjdk-11-hotspot-amd64 && java -version
-
-      - run: ./gradlew -i clean build --stacktrace
+      - run: mkdir ~/tarballs
```
Comment (michaelsembwever) [inline .circleci/config.yml:107]: please can this `mkdir` be put inside gradle?

In hope that CI configurations are as thin as possible, so it is as simple as possible to repeat these steps outside of CircleCI, whether that be on locally or in https://ci-cassandra.apache.org/job/cassandra%7Esidecar/ 

(trivial here i know, just hoping that we'll see better practices and examples with a new project than what we have with that main project…)

Comment (rustyrazorblade) [reply .circleci/config.yml:107]: Good suggestion, I'll move it in.

File: cassandra-integration-tests/src/test/java/org/apache/cassandra/sidecar/common/StatusTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public class StatusTest
+{
+    private static final Logger logger = LoggerFactory.getLogger(StatusTest.class);
+
+    @BeforeEach
+    void setupData(CassandraTestContext context)
+    {
+        logger.info("setup {}", context.container);
+
```
Comment (dineshjoshi) [inline cassandra-integration-tests/src/test/java/org/apache/cassandra/sidecar/common/StatusTest.java:41]: Nit: Extra newline.

File: cassandra-integration-tests/src/test/java/org/apache/cassandra/sidecar/common/testing/CassandraPod.java [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+        Boolean started = false;
+        String response = "";
+
+        for (int i = 0; i < maxTime; i++)
+        {
+            // we sleep in the beginning because the pod will never be ready right away
+            // sometimes K8 seems to hang in CI as well, so this might be enough to let the pod start
+            try
+            {
+                Thread.sleep(1000);
```
Comment (dineshjoshi) [inline cassandra-integration-tests/src/test/java/org/apache/cassandra/sidecar/common/testing/CassandraPod.java:177]: Potentially log a debug statement right before sleeping?

File: common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    {
+        this.cqlSession = cqlSession;
+        this.versionProvider = provider;
+        this.refreshRate = refreshRate;
+    }
+
+    public synchronized void start()
+    {
+        logger.info("Starting health check");
+        // TODO: maybe unhardcode
```
Comment (dineshjoshi) [inline common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java:77]: Do you want to address this before merging?

File: common/src/main/java/org/apache/cassandra/sidecar/common/CassandraVersion.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    private static final Pattern pattern = Pattern.compile(VERSION_REGEXP);
+    private static final Pattern SNAPSHOT = Pattern.compile("-SNAPSHOT");
+
+    public final int major;
+    public final int minor;
+    public final int patch;
+
+    // not going to use these for equality checks but we might need to know what it is at some point
+    private final String extra;
```
Comment (dineshjoshi) [inline common/src/main/java/org/apache/cassandra/sidecar/common/CassandraVersion.java:52]: What is extra? I presume strings like alpha, beta, etc? If so 4.0.0-alpha4 and 4.0.0-alpha5 are technically not the same releases.

File: common/src/test/java/org/apache/cassandra/sidecar/common/CassandraVersionTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+package org.apache.cassandra.sidecar.common;
+
+import org.junit.jupiter.api.Test;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+/**
+ * Tests that CassandraVersion comparisons work correctly
+ */
+public class CassandraVersionTest
```
Comment (dineshjoshi) [inline common/src/test/java/org/apache/cassandra/sidecar/common/CassandraVersionTest.java:28]: Need to add negative test cases for `CassandraVersion::create(String v)`. You should also add tests to validate that the result of `CassandraVersion::toString` is parseable by `CassandraVersion::create(String v)`.

File: scripts/check-kube.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
Comment (dineshjoshi) [inline scripts/check-kube.sh:1]: Empty script?

**Review by dineshjoshi — CHANGES_REQUESTED** [review]
Looks good to me overall. Just minor nits and clarifications.

File: common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java (hunk 4 above)
Comment (rustyrazorblade) [reply common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java:77]: Ah yes, I originally had the time hard coded, removed it, then forgot to remove the comment.

File: cassandra-integration-tests/src/test/java/org/apache/cassandra/sidecar/common/testing/CassandraPod.java (hunk 3 above)
Comment (rustyrazorblade) [reply cassandra-integration-tests/src/test/java/org/apache/cassandra/sidecar/common/testing/CassandraPod.java:177]: Will do

**Review by dineshjoshi — APPROVED** [review]

**Pull request merged**